mirrord-agent/src/steal.rs

```diff
@@ -0,0 +1,407 @@
+/// -> ORIGINAL_CHAIN
+impl SafeIpTables {
+    pub fn new() -> Result<Self> {
+        let ipt = iptables::new(false).unwrap();
+        let random_string = Alphanumeric.sample_string(&mut rand::thread_rng(), 5);
+        let chain_name = format!("MIRRORD_REDIRECT_{}", random_string);
+        ipt.new_chain(IPTABLES_TABLE_NAME, &chain_name)
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        ipt.append(IPTABLES_TABLE_NAME, &chain_name, "-j RETURN")
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        ipt.append(
+            IPTABLES_TABLE_NAME,
+            "PREROUTING",
+            &format!("-j {}", chain_name),
+        )
+        .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        Ok(Self {
+            inner: ipt,
+            chain_name,
+        })
+    }
+
+    pub fn add_redirect(&mut self, redirected_port: Port, target_port: Port) -> Result<()> {
+        self.inner
+            .insert(
+                IPTABLES_TABLE_NAME,
+                &self.chain_name,
+                &format_redirect_rule(redirected_port, target_port),
+                0,
+            )
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))
+    }
+
+    pub fn remove_redirect(&mut self, redirected_port: Port, target_port: Port) -> Result<()> {
+        self.inner
+            .delete(
+                IPTABLES_TABLE_NAME,
+                &self.chain_name,
+                &format_redirect_rule(redirected_port, target_port),
+            )
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))
+    }
+}
+
+impl Drop for SafeIpTables {
+    fn drop(&mut self) {
+        self.inner
+            .delete(
+                IPTABLES_TABLE_NAME,
+                "PREROUTING",
+                &format!("-j {}", self.chain_name),
+            )
+            .unwrap();
+        self.inner
+            .flush_chain(IPTABLES_TABLE_NAME, &self.chain_name)
+            .unwrap();
+        self.inner
+            .delete_chain(IPTABLES_TABLE_NAME, &self.chain_name)
+            .unwrap();
+    }
+}
+
+pub struct StealWorker {
+    pub sender: Sender<DaemonTcp>,
+    iptables: SafeIpTables,
+    ports: HashSet<Port>,
+    listen_port: Port,
+    write_streams: HashMap<ConnectionID, WriteHalf<TcpStream>>,
+    read_streams: StreamMap<ConnectionID, ReaderStream<ReadHalf<TcpStream>>>,
+    connection_index: u64,
+}
+
+impl StealWorker {
+    pub fn new(sender: Sender<DaemonTcp>, listen_port: Port) -> Result<Self> {
+        Ok(Self {
+            sender,
+            iptables: SafeIpTables::new()?,
+            ports: HashSet::default(),
+            listen_port,
+            write_streams: HashMap::default(),
+            read_streams: StreamMap::default(),
+            connection_index: 0,
+        })
+    }
+
+    pub async fn handle_loop(
+        &mut self,
+        mut rx: Receiver<LayerStealTcp>,
+        mut listener: TcpListener,
+    ) -> Result<()> {
```

Comment:
Is this time for using [`ControlFlow`](https://doc.rust-lang.org/stable/std/ops/enum.ControlFlow.html) or should we leave it for a refactor in the future (I would vote for a future refactor)?

Reply:
not sure why control flow is relevant here though - the loop is inside the function?